### ISSUE TITLE
[Android][Native Module] Cannot resize linear layout children in android native module

### ISSUE THREAD

**Issue by sbaar** [facebook/react-native#9275]
I have modified a linear layout with two children to respond to touch and act like a scroll view.

[Here is a video of the correct behavior in normal java land and it not working in react native](https://drive.google.com/open?id=0Bxl2pkHHa3caUFUteHNmSUFlRnM)
[Here is the bare bones sample RN project from above](https://github.com/sbaar/ResizableLLRN)

By looking at logs, the onTouch and setPrimaryContentWidth methods are being called, and in the larger project that this came from I observed that changing the visibility of the view off then on will reveal the actual state of the view (whether the left view is 'opened' or 'closed').

Calling forceUpdate on the js view does not do anything. Calling invalidate or measure does not do anything. The more complex compound view this custom view comes from exhibits many related problems besides this one, such as visibility of children not visibly changing, recyclerview contents not not appearing to change until touched and scrolled, etc, though I hope solving this current issue will help the others.

**Comment by sbaar** [comment]
[Stackoverflow](http://stackoverflow.com/a/39105650/1767887)
Updating the shadow nodes manually by inserting hooks into the custom view was the only solution.

This is an unreliable solution however. For some updates I had to post delayed runnables so they would measure out correctly. As you can probably tell, I don't know exactly what's going here (which is the benefit or rn? hides complexity?) but please expose a reliable invalidation method in native code.

In View Manager (ignore that first break, github formatting)


```
public LayoutShadowNode createShadowNodeInstance() {
    node = new MyShadowNode();
    return node;
}
protected CustomView createViewInstance(final ThemedReactContext reactContext) {   
        customView.setRnUpdateListener(new CustomView.RNUpdateListener() {

```
        Method m;
        @Override
        public void needsUpdate() {
            Runnable r = new Runnable() {
                @Override
                public void run() {

                    if (m == null) {
                        try {
                            m = CSSNode.class.getDeclaredMethod("markHasNewLayout");
                            m.setAccessible(true);
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }
                    try {
                        if (node != null) {
                            if (node.hasNewLayout()) node.markLayoutSeen();
                            ReactShadowNode parent = node.getParent();
                            while (parent != null) {
                                if (parent.hasNewLayout()) {
                                    try {
                                        m.invoke(parent);
                                    } catch (Exception e) {
                                        e.printStackTrace();
                                    }
                                    parent.markLayoutSeen();
                                }
                                parent = parent.getParent();
                            }
                            node.markUpdated();
                        }
                        Log.d(getName(), "markUpdated");
                    } 
                }

            };
            reactContext.runOnNativeModulesQueueThread(r);
        }
    });`

**Comment by AndreaOrru** [comment]
Hi @sbaar, I'm having a very similar problem, where I cannot add subviews to a LinearLayout after createViewInstance. I'm not sure I understand your solution, what exactly have you changed? Thanks.

**Comment by sbaar** [comment]
[Here](https://github.com/makemoji/Makemoji-React-Native/blob/98d48229ace77517d8f7fdf4d0122b26b60d48de/android/app/src/main/java/com/makemoji/mojilib/ReactMojiInputLayoutManager.java) is what I ended up doing in the project I was working on.

In the constructor of the view manager, by reflection I get the method markHasNewLayout of CSSNode and method resolveShadowNode of UIImplementation. Override the createShadowNodeInstance to create an instance of [MyShadowNode](https://github.com/makemoji/Makemoji-React-Native/blob/2665dcd28e95c8f3c22665da056756f3fa6c863b/android/app/src/main/java/csslayout/MyShadowNode.java).

Take a good long look at setRnUpdateListener in the view manager. I set a hook for the compound view to call whenever it made a child view visible/gone, and when contained recycler views are notifyDataSetChanged. This callback is called in my closed source view, not just in the viewmanager.  It was most accurate to measure the heights of the child views and what I knew they should be (45 dp, 250dp)here rather than rely on getHeight(). I set this height to the shadow node of the view by getting the UIManager with 
reactContext.getNativeModule(UIManagerModule.class);
then invoking getShadowNode with the view id. Then follow the shadow node tree up, marking that it has a new layout, then marking it as seen. Then mark the node of the current view as markUpdated. Run this code with
reactContext.runOnNativeModulesQueueThread(r);
@AndreaOrru

**Comment by AndreaOrru** [comment]
Thanks for taking the time and explaining it to me. I tried to reproduce that code in my app but I don't have anything like setRnUpdateListener in my view...?
@sbaar

**Comment by sbaar** [comment]
I had to add that interface and listener to my view and call it when
visibility changes. Treat it like any other interface you create, like the
ones that are auto generated when you create a fragment

On Sep 25, 2016 7:47 AM, "Andrea Orru" <notifications@github.com> wrote:

Thanks for taking the time and explaining it to me. I tried to reproduce
that code in my app but I don't have anything like setRnUpdateListener in
my view...?
@sbaar <https://github.com/sbaar>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/facebook/react-native/issues/9275#issuecomment-249425710>,
or mute the thread
